test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/datasource/DefaultObjectGenerationHandler.java

```diff
@@ -46,9 +54,33 @@ public String generateFieldName() {
         };
     }
 
+    // UNSIGNED_LONG is excluded because it is mapped as long
+    // and values larger than long fail to parse.
+    private static final Set<FieldType> EXCLUDED_FROM_DYNAMIC_MAPPING = Set.of(FieldType.UNSIGNED_LONG);
+
     @Override
     public DataSourceResponse.FieldTypeGenerator handle(DataSourceRequest.FieldTypeGenerator request) {
-        return new DataSourceResponse.FieldTypeGenerator(() -> randomFrom(FieldType.values()));
+        Supplier<DataSourceResponse.FieldTypeGenerator.FieldTypeInfo> generator = switch (request.dynamicMapping()) {
+            case FORBIDDEN -> () -> new DataSourceResponse.FieldTypeGenerator.FieldTypeInfo(randomFrom(FieldType.values()), false);
+            case FORCED -> () -> {
+                var fieldType = ESTestCase.randomValueOtherThanMany(
+                    EXCLUDED_FROM_DYNAMIC_MAPPING::contains,
+                    () -> ESTestCase.randomFrom(FieldType.values())
+                );
+
+                return new DataSourceResponse.FieldTypeGenerator.FieldTypeInfo(fieldType, true);
+            };
+            case SUPPORTED -> () -> {
+                boolean isDynamic = ESTestCase.randomBoolean();
+
+                var excluded = isDynamic ? EXCLUDED_FROM_DYNAMIC_MAPPING : Set.of();
+                var fieldType = ESTestCase.randomValueOtherThanMany(excluded::contains, () -> ESTestCase.randomFrom(FieldType.values()));
+
+                return new DataSourceResponse.FieldTypeGenerator.FieldTypeInfo(fieldType, isDynamic);
+            };
```

Comment:
Nit: refactor to use the switch to just set the value of `useDynamic, then have two branches for dynamic/non-dynamic to avoid duplication.